yaksh/evaluator_tests/test_python_evaluation.py

```diff
@@ -2,52 +2,115 @@
-                                 "test_id": u'null', 
-                                 "pos_args": ["3", "2"], 
-                                 "kw_args": {}
-                                }]
-        get_class = PythonCodeEvaluator(self.test_case_data, self.test, self.language, user_answer, ref_code_path=None, in_dir=None)
-        result = get_class.evaluate()
+        user_answer = dedent("""
+        def add(a, b):
+            return a - b
+        """)
+        get_evaluator = PythonCodeEvaluator(self.test_case_data, self.test, self.language, user_answer, ref_code_path=None, in_dir=None)
+        result = get_evaluator.evaluate()
         self.assertFalse(result.get("success"))
         self.assertEqual(result.get("error"), "AssertionError  in: assert add(3, 2) == 5")
 
     def test_infinite_loop(self):
-        user_answer = "def add(a, b):\n\twhile True:\n\t\tpass"""
-        test_case_data = [{"func_name": "add", 
-                                 "expected_answer": "5", 
-                                 "test_id": u'null', 
-                                 "pos_args": ["3", "2"], 
-                                 "kw_args": {}
-                                }]
-        get_class = PythonCodeEvaluator(self.test_case_data, self.test, self.language, user_answer, ref_code_path=None, in_dir=None)
-        result = get_class.evaluate()
+        user_answer = dedent("""
+        def add(a, b):
+            while True:
+                pass
+        """)
+        timeout_msg = ("Code took more than {0} seconds to run. "
+                       "You probably have an infinite loop in your code.").format(SERVER_TIMEOUT)
+        get_evaluator = PythonCodeEvaluator(self.test_case_data, self.test, self.language, user_answer, ref_code_path=None, in_dir=None)
+        result = get_evaluator.evaluate()
+        self.assertFalse(result.get("success"))
+        self.assertEquals(result.get("error"), timeout_msg)
+
+    def test_syntax_error(self):
+        user_answer = dedent("""
+        def add(a, b);
+            return a + b
+        """)
+        syntax_error_msg = "SyntaxError"
+        get_evaluator = PythonCodeEvaluator(self.test_case_data, self.test, self.language, user_answer, ref_code_path=None, in_dir=None)
+        result = get_evaluator.evaluate()
+        self.assertFalse(result.get("success"))
+        self.assertIn(syntax_error_msg, result.get("error"))
+
+    def test_indent_error(self):
+        user_answer = dedent("""
+        def add(a, b):
+        return a + b
+        """)
+        indent_error_msg = "IndentationError"
+        get_evaluator = PythonCodeEvaluator(self.test_case_data, self.test, self.language, user_answer, ref_code_path=None, in_dir=None)
+        result = get_evaluator.evaluate()
+        self.assertFalse(result.get("success"))
+        self.assertIn(indent_error_msg, result.get("error"))
+
+    def test_name_error(self):
+        user_answer = ""
+        name_error_msg = "NameError"
```

Comment:
So this test will pass even if the error was simply NameError with no backtrace information!  Similarly for the others.